### PR TITLE
ARM64 Support

### PR DESCRIPTION
**Pull request by hrydgard**
This isn't 100% completed (software skinning in the vertex loader is missing, for example) but it's getting good enough to merge, as it's starting to run really well.

Not for immediate merge, but probably quite soon. Making a pull request now to see what the Travis bot makes of it, although we already know it has the wrong NDK installed..

Oh, how neat, it's 64 commits heh.

### CODE REVIEW

File: Core/MIPS/ARM64/Arm64Jit.cpp [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+//memory regions:
+//
+// 08-0A
+// 48-4A
+// 04-05
+// 44-45
+// mov eax, addrreg
+	// shr eax, 28
+// mov eax, [table+eax]
+// mov dreg, [eax+offreg]
```
Comment (unknownbrackets) [inline Core/MIPS/ARM64/Arm64Jit.cpp]: Probably not relevant anymore?

-[Unknown]

File: GPU/Common/VertexDecoderCommon.cpp [hunk 2]
```diff
@@ -134,7 +134,8 @@ void PrintDecodedVertex(VertexReader &vtx) {
 	printf("P: %f %f %f\n", pos[0], pos[1], pos[2]);
 }
 
-VertexDecoder::VertexDecoder() : jitted_(0), decoded_(nullptr), ptr_(nullptr) {
+VertexDecoder::VertexDecoder() : jitted_(0), decoded_(nullptr), ptr_(nullptr)
+{
```
Comment (unknownbrackets) [inline GPU/Common/VertexDecoderCommon.cpp]: Hmm, this change is probably accidental?

-[Unknown]

File: UI/MiscScreens.cpp [hunk 3]
```diff
@@ -439,6 +439,12 @@ void LogoScreen::render() {
 
 #ifdef _WIN32
 	dc.DrawText(screenManager()->getThin3DContext()->GetInfoString(T3DInfo::APINAME), bounds.centerX(), bounds.y2() - 100, colorAlpha(0xFFFFFFFF, alphaText), ALIGN_CENTER);
+#elif defined(ANDROID)
```
Comment (unknownbrackets) [inline UI/MiscScreens.cpp]: Well, I'd say any ARM64 qualifies as an experimental ARM64 build.  Though not sure what work is required to do iOS.

-[Unknown]

File: android/jni/Application.mk [hunk 4]
```diff
@@ -1,6 +1,7 @@
 APP_STL := gnustl_static
 APP_PLATFORM := android-9
-APP_ABI := armeabi-v7a x86
+#APP_ABI := armeabi-v7a x86
 #APP_ABI := armeabi-v7a
+APP_ABI := arm64-v8a armeabi-v7a
```
Comment (unknownbrackets) [inline android/jni/Application.mk]: Probably want to add back x86 at this point.

-[Unknown]

**Comment by unknownbrackets** [comment]
Nice.  But, you know what that means - you've gotta keep it to 64 now, haha.

-[Unknown]

**Comment by unknownbrackets** [comment]
Travis should work now if you rebase - 1e757af64528279fae877403f55c106add97751e.

-[Unknown]

**Comment by unknownbrackets** [comment]
Hmm, still not always working.  Guess I'll need to do something more robust with the download...

-[Unknown]

**Comment by hrydgard** [comment]
Looks like my cmake change broke blackberry, too. hmm

**Comment by hrydgard** [comment]
@unknownbrackets hm, now it completed the download but then didn't do much..

**Comment by hrydgard** [comment]
Hm, for some reason it has started crashing in a few games.. will need to figure that out before merging.

**Comment by unknownbrackets** [comment]
I've tried tweaking some settings - a2dd635d6c0c9d9ff64857ceed93f0962bd43e78.  Not sure why it's being temperamental...

-[Unknown]

**Comment by hrydgard** [comment]
OK, annoying... Anyway, disabling swl/swr/lwl/lwr took care of the new crashes. Can solve it later I guess...

File: Core/MIPS/ARM64/Arm64CompLoadStore.cpp [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+				return;
+			}
+		}
+
+		u32 iaddr = gpr.IsImm(rs) ? offset + gpr.GetImm(rs) : 0xFFFFFFFF;
+		bool doCheck = false;
+		FixupBranch skip;
+
+		if (gpr.IsImm(rs) && Memory::IsValidAddress(iaddr)) {
+			u32 addr = iaddr & 0x3FFFFFFF;
```
Comment (unknownbrackets) [inline Core/MIPS/ARM64/Arm64CompLoadStore.cpp]: I guess this masking should probably be using a constant, hmm.  Anyway, it's not necessary here in 64-bit right?  I wonder if it's better with or without for SetRegImm...

-[Unknown]

File: Core/MIPS/ARM64/Arm64CompLoadStore.cpp [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+
+			case 38: // lwr
+				LDR(SCRATCH1, MEMBASEREG, SCRATCH1);
+				ANDI2R(gpr.R(rt), gpr.R(rt), 0xffffff00 << (24 - shift), SCRATCH2);
+				ORR(gpr.R(rt), gpr.R(rt), SCRATCH1, ArithOption(gpr.R(rt), ST_LSR, shift));
+				break;
+
+			case 42: // swl
+				LDR(SCRATCH2, MEMBASEREG, SCRATCH1);
+				// Don't worry, can't use temporary.
```
Comment (unknownbrackets) [inline Core/MIPS/ARM64/Arm64CompLoadStore.cpp]: Is this still true (and below) for arm64?  Haven't thought about it much, but I know imms are represented differently.

-[Unknown]

Comment (hrydgard) [inline Core/MIPS/ARM64/Arm64CompLoadStore.cpp]: ARM64 logic imms are able to represent all possible contiguous stretches of 1s, so this shouldn't be a concern. Might be some bug though I guess... Should really pass in INVALID_REG there to be sure.

File: Core/MIPS/ARM64/Arm64CompLoadStore.cpp (hunk 5 above)
Comment (hrydgard) [inline Core/MIPS/ARM64/Arm64CompLoadStore.cpp]: Probably best to just get rid of it.

**Comment by unknownbrackets** [comment]
I wrote this test program:

```c
#include <errno.h>
#include <stdio.h>
#include <stdint.h>
#include <sys/mman.h>

void *Grab4GB() {
	// We want at least 4GB with the lower 32 bits all zero.
	// To make sure we can align, first we ask for 8GB of address space.
	static const size_t wanted = 8 * 1024 * 1024 * 1024ULL;
	static const size_t needed = 4 * 1024 * 1024 * 1024ULL;
	void *basep = mmap(0, wanted, PROT_READ | PROT_WRITE, MAP_NORESERVE | MAP_ANON | MAP_SHARED, -1, 0);
	if (basep == MAP_FAILED) {
		return 0;
	}
	munmap(basep, wanted);

	uintptr_t base = (uintptr_t)basep;
	if ((base & 0xFFFFFFFFULL) != 0) {
		base = (base & ~0xFFFFFFFFULL) + 0x100000000ULL;
		basep = (void *)base;
	}

	return mmap(basep, needed, PROT_READ | PROT_WRITE, MAP_NORESERVE | MAP_ANON | MAP_SHARED, -1, 0);
}

void Release4GB(void *p) {
	static const size_t needed = 4 * 1024 * 1024 * 1024ULL;
	munmap(p, needed);
}

int main(int argc, char *argv[]) {
	void *base1 = Grab4GB();
	void *base2 = Grab4GB();

	if (!base1 || !base2) {
		perror("Darn, failed.");
	} else {
		printf("Base 1: %p\n", base1);
		printf("Base 2: %p\n", base2);
	}

	if (base1) {
		Release4GB(base1);
	}
	if (base2) {
		Release4GB(base2);
	}

	return 0;
}
```

On Linux, this seems (?) to successfully get unique 4GB base addresses that are perfectly usable.  It didn't seem to give me any overlap (this seems like it's only a problem if the first parameter is specified.)

Does this work on Android 64?  Do we really need the precarious base?  Note that MAP_NORESERVE is important here, not sure if Android will have it.  We don't need swap here since we separately map the ranges inside that we actually need.  In fact, I wonder if we should ideally keep this mapping with PROT_NONE...

-[Unknown]

**Comment by hrydgard** [comment]
@unknownbrackets , cool, that looks way better than the precarious base. I'll give it a try soon.

**Comment by unknownbrackets** [comment]
I found this: https://code.google.com/p/nativeclient/issues/detail?id=251

Looks like it won't work on Mac/iOS and possibly SUSE (and is possibly not needed there.)  Here it's only used on Linux:
http://www.opensource.apple.com/source/JavaScriptCore/JavaScriptCore-7534.57.3/wtf/OSAllocatorPosix.cpp?txt

-[Unknown]

File: GPU/Common/VertexDecoderArm64.cpp [hunk 7]
```diff
@@ -0,0 +1,779 @@
...
+
+	// Keep the scale/offset in a few fp registers if we need it.
+	if (prescaleStep) {
+		MOVP2R(X3, &gstate_c.uv);
+		if (cpu_info.bNEON) {
+			fp.LDR(64, INDEX_UNSIGNED, neonUVScaleReg, X3, 0);
+			fp.LDR(64, INDEX_UNSIGNED, neonUVOffsetReg, X3, 8);
+			if ((dec.VertexType() & GE_VTYPE_TC_MASK) == GE_VTYPE_TC_8BIT) {
+				fp.MOVI2FDUP(neonScratchRegD, by128, scratchReg);
+				fp.FMUL(32, neonUVScaleReg, neonUVScaleReg, neonScratchRegD);
```
Comment (unknownbrackets) [inline GPU/Common/VertexDecoderArm64.cpp]: I suppose if this used SCONV it would get rid of the constants entirely?

-[Unknown]

**Comment by hrydgard** [comment]
Can be further optimized, but good enough for now, so let's merge.

**Pull request merged**